### DISCUSSION TITLE
`nx serve` works strangely with Webpack 5 and ESM

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/nest-commander#922] [Q&A]
I want to build an app using Nx + Nest + Webpack 5 + TS + ESM + Nest Commander, but I have the following issue.

[Reproduction](https://gitlab.com/r10n/nest-commander) steps:

0. [`pnpm i`] Install the dependencies.

1. [`pnpm serve:gnew:dev` or `pnpm serve:gnew`] `nx serve` does not work unless run with `--inspect false` and `-r` option being processed by (Nest) Commander.

```bash
$ pnpm serve:gnew:dev

> @gnew/source@0.0.0 serve:gnew:dev $repo_root
> nx run gnew:serve -c development --inspect false --watch false


> nx run gnew:serve:development --inspect false --watch false

command : {
  inputs: [],
  options: {
    rrr: '$repo_root/node_modules/.pnpm/source-map-support@0.5.19/node_modules/source-map-support/register.js'
  }
}
{ stdAll: 'test', exitCode: 0 }

 —————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Successfully ran target serve for project gnew
```

2. [`pnpm serve:gnew:dev:debug` or `pnpm serve:gnew:debug`] When run with `--inspect true`, it is successful, but the app is not executed (at least no app logs are output).

```bash
$ pnpm serve:gnew:debug

> @gnew/source@0.0.0 serve:gnew:debug $repo_root
> nx run gnew:serve --inspect true --host 0.0.0.0 --port 9221 --watch false


> nx run gnew:serve:development --inspect true --host 0.0.0.0 --port 9221 --watch false


 —————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Successfully ran target serve for project gnew
```

3. [`node dist/apps/gnew/main.mjs`] It works though when the app is built and executed using `node`.

```bash
$ node dist/apps/gnew/main.mjs
command : { inputs: [], options: {} }
{ stdAll: 'test', exitCode: 0 }
```

4. [`pnpm serve:gnew:dev` or `pnpm serve:gnew`] When I remove `-r` option (or comment out the `@Option()` decorator along with the `parse*()` method), `nx serve --inspect false` fails to output the logs too, though Nx says it was run successfully. Building and running the app using `node` works without issues.

```bash
$ pnpm serve:gnew:dev

> @gnew/source@0.0.0 serve:gnew:dev $repo_root
> nx run gnew:serve -c development --inspect false --watch false


> nx run gnew:serve:development --inspect false --watch false


 —————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Successfully ran target serve for project gnew
```

**Comment by jmcdo29** [comment]
I believe this would have to do with how `nx` is passing `node -r` to register the `source-map-support/register.js` package so that the source maps are properly respected. This is one of those weird moments where essentially you'd be passing something like `node -r source-map-support/register dist/apps/gnew/main.mjs -r some-other-option` so commander _could_ be seeing two `-r`'s. I believe there should be a way to separate them out between the `node` command and your command, but I'll probably need to tinker with it a bit

**Reply by tukusejssirs** [reply]
Well, in this case (at least for now), I have no need to add `-r` option to my app using Nest Commander, however, without it, `nx serve` fails to run it.

Actually, there are to problems:

1. `pnpm nx serve gnew` fails running with `error: unknown option '--inspect=localhost:9229'`;
2. `pnpm nx serve gnew --inspect` fails running with `error: unknown option '-r'` when `-r` option is not processed by Nest Commander.

It might be because of some Nx bug, but I am not that good dev to be able to tell that with certainty.

**Reply by jmcdo29** [reply]
> Well, in this case (at least for now), I have no need to add -r option to my app using Nest Commander, however, without it, nx serve fails to run it.

Whether you have a need for it or not, Nx will add it by default because that's how Nx works. 

I'll clone your reproduction and take a look though, see if I can figure out what's being parsed where

**Reply by jmcdo29** [reply]
Okay, this command makes it work as you expect it to: `pnpm nx serve gnew --inspect=false --runtimeArgs="-r from-command-line"`. The `--runtimeArgs` gets passed to the `process.argv` in the position after the `node` command has ingested the `-r` for the inline require. Also you can use `pnpm nx serve gnew --inspect=false --runtimeArgs="-rrr=from-command-line"`

**Reply by jmcdo29** [reply]
You can also set [`inspect: false` in the `project.json`](https://nx.dev/packages/node/executors/node#inspect) so you don't need the `--inspect=false` every time

**Reply by tukusejssirs** [reply]
> Okay, this command makes it work as you expect it to: `pnpm nx serve gnew --inspect=false --runtimeArgs="-r from-command-line"`. The `--runtimeArgs` gets passed to the `process.argv` in the position after the `node` command has ingested the `-r` for the inline require. Also you can use `pnpm nx serve gnew --inspect=false --runtimeArgs="-rrr=from-command-line"`

Well, it does not work for me with `nx@16.2.1` and `pnpm@8.3.1`:

```bash
$ pnpm nx serve gnew --inspect=false --runtimeArgs="-r"

> nx run gnew:serve:development --inspect=false --runtimeArgs=-r

error: unknown option '-r'
```

> You can also set [inspect: false in the project.json](https://nx.dev/packages/node/executors/node#inspect) so you don't need the --inspect=false every time

I know that, but I actually want to be able to debug the app using my IDE, therefore it is strange it fails. :thinking:

**Reply by tukusejssirs** [reply]
`pnpm nx serve gnew --inspect=false --runtimeArgs='-r from-command-line' --watch false` and `pnpm nx serve gnew --inspect=false --runtimeArgs='-r' --watch false` work though. :thinking:

**Reply by tukusejssirs** [reply]
And requirement to use `--inspect false` might not be an issue, as when I use `--inspect true --host 0.0.0.0 --port 9221`, it seems to work as expected.

Therefore, I have not idea why `pnpm nx serve gnew --runtimeArgs='-r from-command-line'` works for you, but not for me. :thinking:

Anyway, thank you for your time and helping me out (again)! :pray:

**Reply by jmcdo29** [reply]
> `pnpm nx serve gnew --runtimeArgs='-r from-command-line'`

This line on its own _didn't_ work. The `--inspect` gets passed on and causes an error. I would say this is most likely an issue with how Nx is calling `node` programatically and how `commander` is parsing the arguments. That's definitely an error in Nx's format (nest-commander doesn't bold text by default) so it's something within their `@nx/node:node` executor.

**Reply by tukusejssirs** [reply]
> This line on its own didn't work. The --inspect gets passed on and causes an error. I would say this is most likely an issue with how Nx is calling node programatically and how commander is parsing the arguments. That's definitely an error in Nx's format (nest-commander doesn't bold text by default) so it's something within their @nx/node:node executor.

Sorry for confusing you. Yes, I have tried it with and without `--inspect false` (but copied the command without it), all failed, until I added `--watch false`. I have already shared the output in [this comment](https://github.com/jmcdo29/nest-commander/discussions/922#discussioncomment-5961504), but here you go again:

```bash
$ pnpm nx serve gnew --inspect=false --runtimeArgs="-r"

> nx run gnew:serve:development --inspect=false --runtimeArgs=-r

error: unknown option '-r'


>  NX  Process exited with code 1, waiting for changes to restart...
```

```bash
$ pnpm nx serve gnew --inspect=false --runtimeArgs="-r from-command-line"

> nx run gnew:serve:development --inspect=false --runtimeArgs=-r from-command-line

error: unknown option '-r'


>  NX  Process exited with code 1, waiting for changes to restart...
```

**Reply by jmcdo29** [reply]
Go figure :shrug: Might be a difference of shells? Or maybe how the OS is interpreting things and sending them on? Really hard to say. I still think it's an issue with `nx` more than `nest-commander` as `nest-commander` doesn't do command line arg validation
```sh
pnpm nx serve gnew --inspect=false --runtimeArgs="-r from-command-line"
# Found existing alias for "pnpm". You should use: "pn"

> nx run gnew:serve:development --inspect=false --runtimeArgs=-r from-command-line

command : { inputs: [], options: { rrr: ' from-command-line' } }
{ stdAll: 'test', exitCode: 0 }

>  NX  Process exited with code 0, waiting for changes to restart...
```

OS: Arch Linux 6.3.2
Shell: zsh 5.9
Node: 18.16.0
Package Manager: pnpm 8.4.0
Nx: 16.2.1

**Reply by tukusejssirs** [reply]
Our OS and versions are quick similar:

| Type            | Name       | version         | Is different?        |
|-----------------|------------|-----------------|----------------------|
| OS              | Arch Linux | rolling         | no                   |
| Kernel          | Linux      | `6.3.2-arch1-1` | no                   |
| Shell           | BASH       | `5.1.16`        | yes                  |
|                 | Node       | `20.2.0`        | yes (mine is newer)  |
| Package Manager | PNPM       | `8.3.1`         | yes (yours is newer) |
|                 | Nx         | `16.2.1`        | no                   |

I’ll upgrade my `pnpm` version, but I think that should not have much to do with this issue. Same for the shell: AFAIK in this case, both ZSH and BASH behave the same way.

> I still think it's an issue with nx more than nest-commander as nest-commander doesn't do command line arg validation

Agreed. So I’ll open an issue there.

---

Thanks again for help! :pray:

**Reply by tukusejssirs** [reply]
I think I understand now the difference in your and my output: you run `pnpm nx run gnew:serve:development --inspect=false --runtimeArgs="-r from-command-line"` with `-r` option defined using `@Option()` decorator, however, I executed it when it was commented out. Therefore, now I know how to run it _when I need to define my own `-r` option value_.

However, when the `-r` option is not present in my app, `pnpm nx run gnew:serve:development --inspect=false --runtimeArgs="-r from-command-line"` will fail just as I have mentioned above.

Still, it seems like an issue with Nx, not with Nest Commander.
